internal/tokenclient/tokenclient_test.go

```diff
@@ -0,0 +1,273 @@
+					Token:               "token value",
+					ExpirationTimestamp: metav1.NewTime(mockTime.Add(25 * time.Minute)),
+				},
+			},
+			expected: expected{
+				token: "token value",
+				ttl:   25 * time.Minute,
+			},
+		},
+		{
+			name:               "returns errors from howToFetchTokenFromAPIServer",
+			expirationSeconds:  444,
+			serviceAccountName: "service-account-name",
+			tokenResponseError: errors.New("has an error"),
+			expected: expected{
+				errMessage: "error creating token: has an error",
+			},
+		},
+		{
+			name:               "errors when howToFetchTokenFromAPIServer returns nil",
+			expirationSeconds:  333,
+			serviceAccountName: "service-account-name",
+			expected: expected{
+				errMessage: "tokenRequest is nil after request",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			mockClock := clocktesting.NewFakeClock(mockTime.Time)
+			var log bytes.Buffer
+
+			require.NotEmpty(t, tt.serviceAccountName)
+
+			mockClient := fake.NewSimpleClientset()
+			tokenClient := New(
+				tt.serviceAccountName,
+				mockClient.CoreV1().ServiceAccounts("any-namespace-works"),
+				nil,
+				plog.TestLogger(t, &log),
+				WithExpirationSeconds(tt.expirationSeconds),
+			)
+			tokenClient.clock = mockClock
+
+			mockClient.PrependReactor(verb, resource, func(action coretesting.Action) (handled bool, ret runtime.Object, err error) {
+				require.Equal(t, tt.serviceAccountName, action.(coretesting.CreateActionImpl).Name)
+
+				tokenRequest := action.(coretesting.CreateAction).GetObject().(*authenticationv1.TokenRequest)
+				require.NotNil(t, tokenRequest)
+				require.Equal(t, tt.expirationSeconds, *tokenRequest.Spec.ExpirationSeconds)
+				require.Empty(t, tokenRequest.Spec.Audiences)
+				require.Empty(t, tokenRequest.Spec.BoundObjectRef)
+
+				return true, tt.tokenResponseValue, tt.tokenResponseError
+			})
+
+			token, ttl, err := tokenClient.fetchToken(context.Background())
+
+			if tt.expected.errMessage != "" {
+				require.ErrorContains(t, err, tt.expected.errMessage)
+			} else {
+				require.Equal(t, tt.expected.token, token)
+				require.Equal(t, tt.expected.ttl, ttl)
+			}
+		})
+	}
+}
+
+func TestStart_HappyPath(t *testing.T) {
+	mockClient := fake.NewSimpleClientset()
+	now := time.Now()
+	var log bytes.Buffer
+
+	type receivedToken struct {
+		token string
+		ttl   time.Duration
+	}
+
+	var receivedTokens []receivedToken
+
+	tokenClient := New(
+		"service-account-name",
+		mockClient.CoreV1().ServiceAccounts("any-namespace-works"),
+		func(token string, ttl time.Duration) {
+			t.Logf("received token %q with ttl %q", token, ttl)
+			receivedTokens = append(receivedTokens, receivedToken{
+				token: token,
+				ttl:   ttl,
+			})
+		},
+		plog.TestLogger(t, &log),
+	)
+
+	type reactionResponse struct {
+		status authenticationv1.TokenRequestStatus
+		err    error
+	}
+
+	var reactionResponses []reactionResponse
+
+	for i := int64(0); i < 1000; i++ {
```

Comment:
Rather than looping it might be cool to manually construct this list. That way, the responses could have different TTLs, some could be errors, etc. Then the assertions at the end could make an assertion about every callback that happened, roughly when they should have happened, etc. Seems like the test only needs a handful, rather than 1000?

Reply:
Maybe by only using a handful then the test runtime time could also be reduced slightly? Although maybe 10 seconds is not that bad when running all tests because it should get parallelized with other packages, I think.

Reply:
10 seconds with a token lifetime of 50ms loops around 200 times, which is probably a lot more than we need for this style of testing. We could tune both of those down quite a bit and then introduce different scenarios in a test table:

Scenario 1: valid token, valid token, valid token.... valid token
Scenario 2: error, error, error, valid token
Scenario 3: valid token, error, valid token, error

etc